src/Compilers/CSharp/Portable/CSharpParseOptions.cs

```diff
@@ -186,19 +187,49 @@ internal ImmutableArray<ImmutableArray<string>> InterceptorsPreviewNamespaces
                 }
 
                 // e.g. [["System", "Threading"], ["System", "Collections"]]
-                ImmutableArray<ImmutableArray<string>> previewNamespaces;
-                if (Features.TryGetValue("InterceptorsPreviewNamespaces", out var namespaces))
+                ImmutableArray<ImmutableArray<string>> previewNamespaces = Features.TryGetValue("InterceptorsPreviewNamespaces", out var namespaces) && namespaces.Length > 0
+                    ? makeNamespaces(namespaces)
+                    : ImmutableArray<ImmutableArray<string>>.Empty;
+
+                ImmutableInterlocked.InterlockedInitialize(ref _interceptorsPreviewNamespaces, previewNamespaces);
+                return previewNamespaces;
+
+                ImmutableArray<ImmutableArray<string>> makeNamespaces(string namespaces)
                 {
-                    previewNamespaces = namespaces
-                        .Split(';')
-                        .SelectAsArray(segment => segment.Split('.').ToImmutableArray());
+                    var builder = ArrayBuilder<ImmutableArray<string>>.GetInstance();
+                    var singleNamespaceBuilder = ArrayBuilder<string>.GetInstance();
+                    int currentIndex = 0;
+                    while (currentIndex < namespaces.Length && namespaces.IndexOf(';', currentIndex) is not -1 and var semicolonIndex)
+                    {
+                        var namespaceParts = makeSingleNamespaceParts(singleNamespaceBuilder, namespaces.AsSpan(currentIndex, semicolonIndex - currentIndex));
+                        if (!builder.Any(ns => ns.SequenceEqual(namespaceParts)))
+                        {
+                            builder.Add(namespaceParts);
+                        }
```

Comment:
Consider adding this `if` statement to `makeSingleNamespaceParts()` (perhaps renamed to `addSingleNamespace()`), to avoid duplicating here and below.